### PR TITLE
re-implement beam search on top of vllm core

### PR DESCRIPTION
**Pull request by youkaichao**
submitting the pr draft on behalf of @lunreclipse .

### CODE REVIEW

**Comment by zhuohan123** [comment]
Have you measured the performance?

**Comment by youkaichao** [comment]
> Have you measured the performance?

not yet, waiting for gpus to test.

**Comment by youkaichao** [comment]
the command I use to benchmark:

`python benchmarks/benchmark_throughput.py --input-len 256 --output-len 256 --model meta-llama/Meta-Llama-3-8B --use-beam-search --n 4 --num-prompts 100`

the machine is H100 DGX, and I use one GPU to run the benchmark.

# main branch:

`Throughput: 0.61 requests/s, 310.62 tokens/s`

# this branch
turn on `use_new_beam_search = True` in the benchmark script

`Throughput: 0.10 requests/s, 51.55 tokens/s`

# this branch + prefix caching

add `--enable-prefix-caching --use-v2-block-manager`

`python benchmarks/benchmark_throughput.py --input-len 256 --output-len 256 --model meta-llama/Meta-Llama-3-8B --use-beam-search --n 4 --num-prompts 100 --enable-prefix-caching --use-v2-block-manager`

`Throughput: 1.08 requests/s, 551.83 tokens/s`

# this branch + prefix caching (with block size 1)

flash attention and xformers do not support block size 1.

it looks pretty exciting that using this branch and prefix caching is much faster than the main branch.

**Comment by youkaichao** [comment]
cc @WoosukKwon @zhuohan123 @simon-mo  for final decision.

and cc @hrsmanian @zhouyuan @lanking520 @nightflight-dk @HeegonJin @SemMulder @darabos @DhruvaBansal00 @tmostak @physicsrob @YooSungHyun @denadai2 @sjmielke @Reichenbachian @AaronFriel @hinnefe2 @mflaxman10 for community feedback ( and especially community contribution!)

**Review by zhuohan123 — APPROVED** [review]
LGTM! The code looks much cleaner than the code in the current code base! Happy to see this working.

File: vllm/entrypoints/llm.py [hunk 1]
```diff
@@ -352,6 +385,99 @@ def generate(
...
+
+        Args:
+            prompts: A list of prompts. Each prompt can be a string or a list
+                of token IDs.
+            beam_width: The number of beams to keep at each step.
+            max_tokens: The max number of tokens to generate for each prompt.
+        """
+
+        tokenizer = self.get_tokenizer()
+        beam_search_params = SamplingParams(logprobs=beam_width,
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/entrypoints/llm.py:406]: I think at most 1 sample can be eos. any possibility that multiple samples can be eos?

File: benchmarks/benchmark_throughput.py [hunk 2]
```diff
@@ -132,9 +132,23 @@ def run_vllm(
                 max_tokens=output_len,
             ))
 
-    start = time.perf_counter()
-    llm.generate(prompts, sampling_params, use_tqdm=True)
-    end = time.perf_counter()
+    use_new_beam_search = False
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply benchmarks/benchmark_throughput.py:135]: sure, let me expose this as an argument.

File: benchmarks/benchmark_throughput.py [hunk 3]
```diff
@@ -132,9 +132,23 @@ def run_vllm(
...
+        assert use_beam_search
+        prompts = [prompt for prompt, _, _ in requests]
+        # output_len should be the same for all requests.
+        output_len = requests[0][2]
+        start = time.perf_counter()
+        llm.beam_search(prompts,
+                        beam_width=n,
+                        max_tokens=output_len,
+                        ignore_eos=True)
+        end = time.perf_counter()
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply benchmarks/benchmark_throughput.py:151]: I think we can use the current one, but with several new lines of code.

File: vllm/entrypoints/llm.py [hunk 4]
```diff
@@ -352,6 +385,99 @@ def generate(
...
+                    if result.outputs[0].logprobs is not None:
+                        # if `result.outputs[0].logprobs` is None, it means
+                        # the sequence is completed because of the max-model-len
+                        # or abortion. we don't need to add it to the new beams.
+                        logprobs = result.outputs[0].logprobs[0]
+                        for token_id, logprob_obj in logprobs.items():
+                            new_beam = BeamSearchSequence(
+                                tokens=current_beam.tokens + [token_id],
+                                cum_logprob=current_beam.cum_logprob +
+                                logprob_obj.logprob)
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/entrypoints/llm.py:455]: no. how does beam search work with length penalty?

File: vllm/entrypoints/llm.py [hunk 5]
```diff
@@ -352,6 +385,99 @@ def generate(
...
+
+                            if token_id == tokenizer.eos_token_id and \
+                                not ignore_eos:
+                                instance.completed.append(new_beam)
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/entrypoints/llm.py:459]: no, how do they work with beam search?

[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/entrypoints/llm.py:465]: fixed in [9504bb4](https://github.com/vllm-project/vllm/pull/8726/commits/9504bb429ae2f07cb2edb1eeeaf3a03f033dcd46)

File: vllm/entrypoints/llm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/entrypoints/llm.py:461]: added TODO in [9504bb4](https://github.com/vllm-project/vllm/pull/8726/commits/9504bb429ae2f07cb2edb1eeeaf3a03f033dcd46)

File: vllm/entrypoints/llm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/entrypoints/llm.py]: changed in [9504bb4](https://github.com/vllm-project/vllm/pull/8726/commits/9504bb429ae2f07cb2edb1eeeaf3a03f033dcd46)

File: benchmarks/benchmark_throughput.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (youkaichao) [reply benchmarks/benchmark_throughput.py]: fixed in [1c22995](https://github.com/vllm-project/vllm/pull/8726/commits/1c22995d33c78851fa74dec23d2437638593fea0)

**Pull request merged**